### PR TITLE
feat: try to proxy body even after body-parser middleware

### LINKED ISSUE #90: Modifying the POST body in onProxyReq
Following this recipe: https://github.com/chimurai/http-proxy-middleware/blob/master/recipes/modify-post.md, I expect to be able conditionally modify the POST body based on the proxy path. The encodeURIComponent portion of this script only works with objects that are one level deep. 

How do I make this work with any object, no matter how many levels deep it might go, or just bypass the onProxyReq method entirely for other URL's I don't care about?
### Setup
- body-parser: 1.15.1
- http-proxy-middleware: 0.15.2
- express: 4.13.4
#### proxy middleware configuration

``` javascript
const apiProxy = proxyMiddleware('/services', {
  target: config.serverUrl,
  changeOrigin: true,
  pathRewrite: {
    '^/services': ''
  },
  onProxyReq: // just like the recipe linked above, but with a conditional check on proxy path
```
#### server mounting

``` javascript
const app = express();
app.use(bodyParser.urlencoded({ extended: true }));
app.use(bodyParser.json());

app.use(apiProxy);
app.listen(3000);
```

### LINKED ISSUE #320: app.use(bodyParser.json()) breaks http-proxy-middleware for HTTP POST JSON with express
For a test case of the following

curl -X POST --header 'Content-Type: application/json' --header 'Accept: text/html' -d '{ "$class": "org.accordproject.cicero.contract.AccordContractState", "stateId": "test2" }' 'http://domain/path'

Where http://domain/path uses http-proxy-middleware with express then the introduction of app.use(bodyParser.json()) for express with this http-proxy-middleware, all together the app.use(bodyParser.json()) breaks this.

This is an unexpected surprise and finding this cause of this error was difficult.

### LINKED ISSUE #417: Requests with a body don't work
### Is this a bug report?

Yes

### Steps to reproduce

1. Setup the project as written in the readme
2. Create an endpoint to check if the request succeeds (mine is on https://webhook.site)
3. Execute a request with no body (using postman) => succeeds
4. Execute the same request with a body (using postman) => timeout/ no response

**The request:**
`http://127.0.0.1/proxy/MY-WEBHOOK-IDENTIFIER`

### Expected behavior

forward `http://127.0.0.1/proxy/MY-WEBHOOK-IDENTIFIER` to `https://webhook.site/MY-WEBHOOK-IDENTIFIER` no matter what content:

No timeout when requesting on `http://127.0.0.1/proxy/MY-WEBHOOK-IDENTIFIER` with a content
No timeout when requesting on `http://127.0.0.1/proxy/MY-WEBHOOK-IDENTIFIER` with no content
No timeout when requesting on `https://webhook.site/MY-WEBHOOK-IDENTIFIER` with or without content

### Actual behavior

Timeout when requesting on `http://127.0.0.1/proxy/MY-WEBHOOK-IDENTIFIER` with a content
No timeout when requesting on `http://127.0.0.1/proxy/MY-WEBHOOK-IDENTIFIER` with no content
No timeout when requesting on `https://webhook.site/MY-WEBHOOK-IDENTIFIER` with or without content

### Setup

`app.use("/proxy", createProxyMiddleware({ target: "https://webhook.site", pathRewrite: { ["^/proxy"]: "" }, logLevel: "debug", changeOrigin: true }));`

- http-proxy-middleware: 1.0.3
- http-proxy-middleware configuration: logLevel=debug, changeOrigin=true, pathRewrite={ ["^/proxy"]: "" }
- server: express + 4.16.3

#### client info

OS: Linux Mint 19.3 "Tricia"
User-Agent: PostmanRuntime/7.24.0

#### target server info

Webhook.site: A free web server to mirror incoming requests

### PR DESCRIPTION
**Pull request by midgleyc**
If a body parser middleware is added before the proxy is configured, proxying POST requests with bodies fails. This fixes that and provides a test, as requested by @chimurai in #460.

Code from #460 by @sarumont and from #320 by @repl-chris used -- thanks to both!

Fixes #90
Fixes #320
Fixes #417

### CODE REVIEW

**Comment by dquang-tran** [comment]
Hi, any updates on this PR? Thanks for doing this.

**Comment by chimurai** [comment]
Thanks @midgleyc for this PR. (Sorry for late reply)

Could you separate `this.fixBody` function to external handler? Similar to `responseInterceptor`'s implementation. https://github.com/chimurai/http-proxy-middleware/tree/master/src/handlers

This'll make `fixBody` optional, since not everyone is having this issue. And it'll prevent bloat in the core functionality.

It's usage will be like this:

```js
const { createProxyMiddleware, fixBody } = require('http-proxy-middleware');

const proxy = createProxyMiddleware({
  /**
   * Fix bodyParser
   **/
  onProxyReq: fixBody,
});
```

Planning to work on a `preset`-like functionality where something like `recommended` contains a collection of handlers for the majority of use-cases.

**Comment by midgleyc** [comment]
@chimurai Done, how does it look now?

File: test/e2e/_utils.ts [hunk 1]
```diff
@@ -1,11 +1,11 @@
 import * as express from 'express';
 import { Express, RequestHandler } from 'express';
 
-export { createProxyMiddleware, responseInterceptor } from '../../dist/index';
+export { createProxyMiddleware, responseInterceptor, fixRequestBody } from '../../dist/index';
 
-export function createApp(middleware: RequestHandler): Express {
+export function createApp(...middleware: RequestHandler[]): Express {
```
Comment (chimurai) [inline test/e2e/_utils.ts:6]: Could you pluralise this to: `...middlewares` ?

**Review by chimurai — CHANGES_REQUESTED** [review]
That's quick  :)

Could you add some documentation before of after:
https://github.com/chimurai/http-proxy-middleware/blob/master/README.md#intercept-and-manipulate-responses

File: src/handlers/fix-request-body.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+ * Fix proxied body if bodyParser is involved.
+ */
+export function fixRequestBody(proxyReq: ClientRequest, req: Request): void {
+  if (!req.body || !Object.keys(req.body).length) {
+    return;
+  }
+
+  const contentType = proxyReq.getHeader('Content-Type') as string;
+  const writeBody = (bodyData: string) => {
+    proxyReq.setHeader('Content-Length', Buffer.byteLength(bodyData));
```
Comment (chimurai) [inline src/handlers/fix-request-body.ts:15]: Can you add this comment to silence DeepCode bot?
```
    // deepcode ignore ContentLengthInCode: bodyParser fix
    proxyReq.setHeader('Content-Length', Buffer.byteLength(bodyData));
```

**Comment by midgleyc** [comment]
Changes made.

You should be able to make changes to my branch directly, if you'd like.

**Review by chimurai — APPROVED** [review]
lgtm

**Comment by chimurai** [comment]
Thanks @midgleyc !

**Comment by midgleyc** [comment]
You're welcome, thanks for merging :)

**Comment by dhurtrci** [comment]
This workaround doesn't seem to work anymore.
In later versions, the proxyReq etc. are set inside the on property:
```

on: {
                        /**
                        * Fix bodyParser
                        **/
                        proxyReq: fixRequestBody
                    }
```

In such versions, I get an error in the fixRequestBody method (where it sets the content type header), and is says (something like) 'cannot change headers after they are sent to the client'.

**Comment by chimurai** [comment]
@dhurtrci Can you please open a new issue and provide a reproduction so it can be investigated?

**Pull request merged**